File: tif_customization/tif_customization/page/field_staff_report/field_staff_report.js
frappe.pages["field-staff-report"].on_page_load = function (wrapper) {
	frappe.require("/assets/tif_customization/css/field_staff_report.css");
	frappe.require("/assets/tif_customization/js/field_visit_drilldown.js", () => {
		new FieldStaffReportPage(wrapper);
	});
};

class FieldStaffReportPage {
	constructor(wrapper) {
		this.suspend_filter_change = false;
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Field Staff Report"),
			single_column: true
		});

		this.setup_layout();
		this.setup_filters();
		this.apply_route_options();
		this.bind_actions();
		this.offset = 0;
		this.page_size = 100;
		this.load_data();
	}

	apply_route_options() {
		const route = frappe.route_options || {};
		if (route.from_date && this.filters.from_date) this.filters.from_date.set_value(route.from_date);
		if (route.to_date && this.filters.to_date) this.filters.to_date.set_value(route.to_date);
		if (route.user && this.filters.user) this.filters.user.set_value(route.user);
		if (route.type && this.filters.type) this.filters.type.set_value(route.type);
		frappe.route_options = null;
	}

	get_month_start() {
		const dt = new Date();
		dt.setDate(1);
		return dt.toISOString().slice(0, 10);
	}

	setup_layout() {
		this.body = $(`
			<div class="p-2">
				<div class="border rounded p-3 mb-3">
					<div class="d-flex align-items-center justify-content-between mb-2">
						<h5 class="mb-0">${__("Filters")}</h5>
						<div>
							<button class="btn btn-sm btn-primary fsr-apply">${__("Apply")}</button>
							<button class="btn btn-sm btn-default fsr-reset">${__("Reset")}</button>
						</div>
					</div>
					<div class="row fsr-filter-grid"></div>
				</div>

				<div class="fsr-kpis mb-2"></div>
				<p class="fsr-breakdown fsr-breakdown-note"></p>

				<div class="border rounded p-3">
					<div class="d-flex align-items-center justify-content-between mb-2">
						<h5 class="mb-0">${__("Field Visit Reporting")}</h5>
						<div>
							<div class="btn-group">
								<button class="btn btn-sm btn-success dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
									${__("Export")}
								</button>
								<div class="dropdown-menu dropdown-menu-right">
									<a class="dropdown-item fsr-export-excel" href="#">${__("Export Excel")}</a>
									<a class="dropdown-item fsr-export-csv" href="#">${__("Export CSV")}</a>
								</div>
							</div>
							<span class="text-muted fsr-count ml-2"></span>
						</div>
					</div>
					<div class="fsr-pager mb-2 d-flex align-items-center"></div>
					<div class="fsr-table"></div>
				</div>
			</div>
		`);

		$(this.page.body).empty().append(this.body);
	}

	make_filter(df) {
		const col = $('<div class="col-md-3 mb-2"></div>').appendTo(this.body.find(".fsr-filter-grid"));
		const parent = $('<div></div>').appendTo(col);
		const control = frappe.ui.form.make_control({
			parent,
			df: { ...df },
			render_input: true
		});
		control.refresh();
		if (df.default) control.set_value(df.default);
		return control;
	}

	setup_filters() {
		this.filters = {
			from_date: this.make_filter({
				label: __("Visit From Date"),
				fieldname: "from_date",
				fieldtype: "Date",
				default: this.get_month_start()
			}),
			to_date: this.make_filter({
				label: __("Visit To Date"),
				fieldname: "to_date",
				fieldtype: "Date",
				default: frappe.datetime.get_today()
			}),
			type: this.make_filter({
				label: __("Type"),
				fieldname: "type",
				fieldtype: "Select",
				options: "\nMarketing\nM&E\nTraining\nMeeting\nAcademic / Other Official Tasks"
			}),
			user: this.make_filter({
				label: __("Field Staff"),
				fieldname: "user",
				fieldtype: "Select",
				options: "\n",
			}),
			province: this.make_filter({
				label: __("Province"),
				fieldname: "province",
				fieldtype: "Select",
				options:
					"\nPunjab\nSindh\nKhyber Pakhtunkhwa\nBalochistan\nAzad Jammu & Kashmir\nGilgit-Baltistan\nIslamabad Capital Territory"
			})
		};
		this.load_team_filter_options();
	}

	async load_team_filter_options() {
		try {
			const r = await frappe.call({
				method: "tif_customization.tif_customization.field_visit_permissions.get_my_field_team",
			});
			const team = r.message?.team || [];
			const opts = [""].concat(
				team.map((t) => t.employee_name || t.user_id || t.employee).filter(Boolean)
			);
			const unique = [...new Set(opts)];
			this.filters.user.df.options = unique.join("\n");
			this.filters.user.refresh();
		} catch (e) {
			console.warn("Could not load field team filter", e);
		}
	}

	bind_actions() {
		this.body.find(".fsr-apply").on("click", () => {
			this.offset = 0;
			this.load_data();
		});
		this.body.find(".fsr-export-excel").on("click", (e) => {
			e.preventDefault();
			this.export_excel();
		});
		this.body.find(".fsr-export-csv").on("click", (e) => {
			e.preventDefault();
			this.export_csv();
		});
		this.body.find(".fsr-reset").on("click", () => {
			this.set_filters({
				from_date: this.get_month_start(),
				to_date: frappe.datetime.get_today(),
				type: "",
				user: "",
				province: ""
			});
			this.offset = 0;
			this.load_data();
		});
		this.body.find(".fsr-pager").on("click", ".fsr-prev:not(:disabled)", () => {
			this.offset = Math.max(0, this.offset - this.page_size);
			this.load_data();
		});
		this.body.find(".fsr-pager").on("click", ".fsr-next:not(:disabled)", () => {
			this.offset = this.offset + this.page_size;
			this.load_data();
		});
		this.body.on("click", "[data-visit-metric]", (e) => {
			const metric = $(e.currentTarget).attr("data-visit-metric");
			if (!metric) return;
			const f = this.get_filter_values();
			frappe.tif_customization.open_visit_drilldown({
				from_date: f.from_date,
				to_date: f.to_date,
				staff: f.user,
				metric,
			});
		});
	}

	on_filter_change() {
		if (this.suspend_filter_change) return;
		this.load_data();
	}

	set_filters(values) {
		this.suspend_filter_change = true;
		Object.entries(values).forEach(([key, value]) => {
			if (this.filters[key]) this.filters[key].set_value(value || "");
		});
		this.suspend_filter_change = false;
	}

	get_filter_values() {
		return {
			from_date: this.filters.from_date.get_value(),
			to_date: this.filters.to_date.get_value(),
			type: this.filters.type.get_value(),
			user: this.filters.user.get_value(),
			province: this.filters.province.get_value()
		};
	}

	load_data() {
		this.body.find(".fsr-table").html(`<div class="text-muted">${__("Loading...")}</div>`);
		const filters = {
			...this.get_filter_values(),
			limit: this.page_size,
			offset: this.offset,
		};
		frappe.call({
			method: "tif_customization.tif_customization.page.field_staff_report.field_staff_report.get_report_data",
			args: { filters },
			freeze: false,
			callback: (r) => {
				const data = r.message || { rows: [], columns: [], labels: {}, total_count: 0, summary: {} };
				this.render_kpis(data.summary || {});
				this.render_breakdown(data.summary || {});
				this.render_table(data.rows || [], data.columns || [], data.labels || {});
				this.render_pager(data);
				const total = data.total_count || 0;
				const shown = data.shown_count || (data.rows || []).length;
				this.body.find(".fsr-count").text(
					total > shown
						? `${shown} ${__("of")} ${total} ${__("records")}`
						: `${total} ${__("records")}`
				);
			},
			error: () => {
				this.body.find(".fsr-table").html(`<div class="text-danger">${__("Failed to load report.")}</div>`);
			}
		});
	}

	build_client_summary(rows) {
		const typeKeys = {
			Marketing: "marketing",
			"M&E": "me",
			Training: "training",
			Meeting: "meeting",
			Other: "other"
		};
		const typeCounts = { marketing: 0, me: 0, training: 0, meeting: 0, other: 0 };
		const staffMap = new Map();

		rows.forEach((row) => {
			const visitType = row.type || "Other";
			const typeKey = typeKeys[visitType] || "other";
			typeCounts[typeKey] += 1;
			const staff = this.get_row_staff(row).trim() || __("Unassigned");
			const staffKey = staff.toLocaleLowerCase();
			if (!staffMap.has(staffKey)) {
				staffMap.set(staffKey, {
					staff,
					total_visits: 0,
					marketing: 0,
					me: 0,
					training: 0,
					meeting: 0,
					other: 0
				});
			}
			const staffRow = staffMap.get(staffKey);
			staffRow.total_visits += 1;
			staffRow[typeKey] += 1;
		});

		const totalVisits = rows.length;
		const staffWise = [...staffMap.values()]
			.map((row) => ({
				...row,
				ratio: totalVisits ? Number((row.total_visits / totalVisits * 100).toFixed(1)) : 0
			}))
			.sort((a, b) => b.total_visits - a.total_visits || a.staff.localeCompare(b.staff));

		return {
			summary: {
				total_visits: totalVisits,
				marketing_visits: typeCounts.marketing,
				me_visits: typeCounts.me,
				training_visits: typeCounts.training,
				meeting_visits: typeCounts.meeting,
				other_visits: typeCounts.other,
				active_staff: staffWise.length,
				visits_per_staff: staffWise.length
					? Number((totalVisits / staffWise.length).toFixed(1))
					: 0
			},
			staff_wise: staffWise
		};
	}

	get_row_staff(row) {
		if (row.type === "Marketing") return row.visit_by || row.owner || "";
		if (row.type === "M&E") return row.me_visit_by || row.owner || "";
		if (row.type === "Training") return row.training_entry_filled_by || row.owner || "";
		if (row.type === "Meeting") return row.mt_visit_by || row.owner || "";
		return row.owner || "";
	}

	render_kpis(summary) {
		const cards = [
			[__("Total Visits"), summary.total_visits || 0, "primary", "visits"],
			[__("Marketing Visits"), summary.marketing_visits || 0, "marketing", "marketing"],
			[__("M&E Visits"), summary.me_visits || 0, "me", "me"],
			[__("Training Visits"), summary.training_visits || 0, "training", "training"],
			[__("Meeting Visits"), summary.meeting_visits || 0, "meeting", "meeting"],
			[__("Other Visits"), summary.other_visits || 0, "other", "other"],
			[__("Active Field Staff"), summary.active_staff || 0, "staff", ""],
		];

		this.body.find(".fsr-kpis").html(
			cards
				.map(
					([label, value, style, metric]) => `
				<div class="fsr-kpi fsr-kpi--${style}" ${metric ? `data-visit-metric="${metric}"` : ""}>
					<div class="fsr-kpi__label">${label}</div>
					<div class="fsr-kpi__value">${frappe.utils.escape_html(String(value))}</div>
					${metric ? `<div class="fsr-kpi__hint">${__("Click to see documents")}</div>` : ""}
				</div>`
				)
				.join("")
		);
	}

	render_breakdown(summary) {
		const total = summary.total_visits || 0;
		const parts = [
			["marketing", __("Marketing"), summary.marketing_visits || 0],
			["me", __("M&E"), summary.me_visits || 0],
			["meeting", __("Meetings"), summary.meeting_visits || 0],
			["training", __("Training"), summary.training_visits || 0],
			["other", __("Other / Academic"), summary.other_visits || 0],
		]
			.map(
				([metric, label, n]) =>
					`<span class="fsr-break-part" data-visit-metric="${metric}">${label} ${n}</span>`
			)
			.join(" + ");
		this.body.find(".fsr-breakdown-note").html(
			`<strong class="fsr-break-part" data-visit-metric="visits">${__("Total visits")}: ${total}</strong>
			= ${parts}.
			${__("This is every Field Visit type. Click any number (cards or this line) to open those documents.")}`
		);
	}

	export_excel() {
		const filters = encodeURIComponent(JSON.stringify(this.get_filter_values()));
		const url = `/api/method/tif_customization.tif_customization.page.field_staff_report.field_staff_report.download_report_excel?filters=${filters}`;
		window.open(url, "_blank");
	}

	export_csv() {
		const filters = encodeURIComponent(JSON.stringify(this.get_filter_values()));
		const url = `/api/method/tif_customization.tif_customization.page.field_staff_report.field_staff_report.download_report_csv?filters=${filters}`;
		window.open(url, "_blank");
	}

	render_table(rows, columns, labels) {
		if (!rows.length) {
			this.body.find(".fsr-table").html(`<div class="text-muted">${__("No data found.")}</div>`);
			return;
		}

		const safeColumns = columns.length ? columns : Object.keys(rows[0] || {});
		const head = safeColumns
			.map((col) => `<th>${frappe.utils.escape_html(labels[col] || col)}</th>`)
			.join("");
		const body = rows
			.map((row) => {
				const tds = safeColumns
					.map((col) => {
						const value = row[col];
						if (col === "name") {
							const id = frappe.utils.escape_html(value || "");
							return `<td><a href="/app/field-visit/${id}">${id}</a></td>`;
						}
						return `<td>${frappe.utils.escape_html(value == null || value === "" ? "-" : String(value))}</td>`;
					})
					.join("");
				return `<tr>${tds}</tr>`;
			})
			.join("");

		this.body.find(".fsr-table").html(`
			<div class="table-responsive">
				<table class="table table-bordered table-hover mb-0">
					<thead><tr>${head}</tr></thead>
					<tbody>${body}</tbody>
				</table>
			</div>
		`);
	}

	render_pager(data) {
		const total = data.total_count || 0;
		const offset = data.offset || 0;
		const shown = data.shown_count || (data.rows || []).length;
		const from = total ? offset + 1 : 0;
		const to = Math.min(offset + shown, total);
		const hasPrev = offset > 0;
		const hasNext = offset + shown < total;
		this.body.find(".fsr-pager").html(`
			<button type="button" class="btn btn-xs btn-default fsr-prev" ${hasPrev ? "" : "disabled"}>${__("Previous")}</button>
			<span class="mx-2 text-muted">
				${__("Showing")} ${from}–${to} ${__("of")} ${total}.
				${__("Click Document No for full details. Use Export for the complete list.")}
			</span>
			<button type="button" class="btn btn-xs btn-default fsr-next" ${hasNext ? "" : "disabled"}>${__("Next")}</button>
		`);
	}
}

File: tif_customization/tif_customization/page/field_staff_report/field_staff_report.py
import json
import csv
import re
from io import StringIO
from datetime import date, datetime, time

import frappe
from frappe import _
from frappe.utils import cint, getdate
from frappe.utils.xlsxutils import make_xlsx

# Screen table only — Field Visit has ~200 columns; dumping them froze the page.
DISPLAY_COLUMNS = [
	"name",
	"visit_date",
	"type",
	"school",
	"officer",
	"category",
	"province",
	"docstatus",
	"owner",
]
DISPLAY_LABELS = {
	"name": "Document No",
	"visit_date": "Visit Date",
	"type": "Type",
	"school": "School / Venue",
	"officer": "Field Staff",
	"category": "Category",
	"province": "Province",
	"docstatus": "Status",
	"owner": "Owner",
}
PAGE_SIZE = 100


@frappe.whitelist()
def get_report_data(filters=None):
	if not frappe.has_permission("Field Visit", "read"):
		frappe.throw(_("You are not permitted to view Field Visit data."))

	filters = _parse_filters(filters)
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	if from_date:
		from_date = getdate(from_date)
	if to_date:
		to_date = getdate(to_date)
	if from_date and to_date and from_date > to_date:
		frappe.throw(_("From Date cannot be after To Date."))

	where_clause, params, visit_date_expr = _build_where(filters, from_date, to_date)
	summary = _sql_summary(where_clause, params)

	for_export = cint(filters.get("for_export"))
	limit_sql = ""
	if not for_export:
		limit = min(max(cint(filters.get("limit") or PAGE_SIZE), 1), 500)
		offset = max(cint(filters.get("offset") or 0), 0)
		params["limit"] = limit
		params["offset"] = offset
		limit_sql = " LIMIT %(limit)s OFFSET %(offset)s"
	else:
		limit = summary["total_visits"]
		offset = 0

	rows = frappe.db.sql(
		f"""
		SELECT
			fv.name,
			fv.type,
			fv.docstatus,
			fv.owner,
			{visit_date_expr} AS visit_date,
			{_school_sql("fv")} AS school,
			{_officer_sql("fv")} AS officer,
			{_category_sql("fv")} AS category,
			{_province_sql("fv")} AS province
		FROM `tabField Visit` fv
		WHERE {where_clause}
		ORDER BY visit_date DESC, fv.modified DESC
		{limit_sql}
		""",
		params,
		as_dict=True,
	)
	_format_status(rows)

	return {
		"columns": DISPLAY_COLUMNS,
		"labels": DISPLAY_LABELS,
		"rows": rows,
		"total_count": summary["total_visits"],
		"shown_count": len(rows),
		"offset": offset,
		"limit": limit if not for_export else len(rows),
		"summary": summary,
		"staff_wise": [],
	}


@frappe.whitelist()
def download_report_excel(filters=None):
	filters = _parse_filters(filters)
	filters["for_export"] = 1
	report = get_report_data(filters=filters)
	columns = report.get("columns", [])
	labels = report.get("labels", {})
	rows = report.get("rows", [])

	xlsx_data = [[labels.get(col, col) for col in columns]]
	for row in rows:
		xlsx_data.append([_excel_value(row.get(col)) for col in columns])

	xlsx_file = make_xlsx(xlsx_data, "Field Staff Report")
	frappe.response["filename"] = "field_staff_report.xlsx"
	frappe.response["filecontent"] = xlsx_file.getvalue()
	frappe.response["type"] = "binary"


@frappe.whitelist()
def download_report_csv(filters=None):
	filters = _parse_filters(filters)
	filters["for_export"] = 1
	report = get_report_data(filters=filters)
	columns = report.get("columns", [])
	labels = report.get("labels", {})
	rows = report.get("rows", [])

	buffer = StringIO()
	writer = csv.writer(buffer)
	writer.writerow([labels.get(col, col) for col in columns])
	for row in rows:
		writer.writerow([_excel_value(row.get(col)) for col in columns])

	frappe.response["filename"] = "field_staff_report.csv"
	frappe.response["filecontent"] = buffer.getvalue()
	frappe.response["type"] = "csv"


def _parse_filters(filters):
	if isinstance(filters, str):
		try:
			return json.loads(filters)
		except Exception:
			return {}
	return filters or {}


def _build_where(filters, from_date, to_date):
	from tif_customization.tif_customization.field_visit_permissions import (
		apply_team_scope_to_conditions,
		can_view_all_field_visits,
		expand_staff_tokens,
		staff_is_in_team,
		staff_match_sql,
		visit_day_sql,
	)

	visit_type = filters.get("type")
	user = filters.get("user")
	province = filters.get("province")

	conditions = ["fv.docstatus < 2"]
	params = {}
	apply_team_scope_to_conditions(conditions, params, alias="fv")
	visit_date_expr = visit_day_sql("fv")

	if from_date:
		conditions.append(f"{visit_date_expr} >= %(from_date)s")
		params["from_date"] = from_date
	if to_date:
		conditions.append(f"{visit_date_expr} <= %(to_date)s")
		params["to_date"] = to_date
	if visit_type:
		conditions.append("fv.type = %(visit_type)s")
		params["visit_type"] = visit_type
	if user:
		if not can_view_all_field_visits() and not staff_is_in_team(user):
			frappe.throw(_("You can only view reports for your field staff."))
		tokens = expand_staff_tokens(user)
		params["staff_tokens"] = tuple(t.lower() for t in tokens) or ("__none__",)
		conditions.append(staff_match_sql("fv", "staff_tokens"))
	if province:
		conditions.append(f"({_province_sql('fv')}) = %(province)s")
		params["province"] = province

	return " AND ".join(conditions), params, visit_date_expr


def _school_sql(alias="fv"):
	a = alias
	return f"""COALESCE(
		NULLIF(TRIM({a}.school_name), ''),
		NULLIF(TRIM({a}.me_school_name), ''),
		NULLIF(TRIM({a}.mt_institute_or_organization_name), ''),
		NULLIF(TRIM({a}.training_venue_name), '')
	)"""


def _officer_sql(alias="fv"):
	a = alias
	return f"""CASE
		WHEN {a}.type = 'Marketing' THEN COALESCE(NULLIF(TRIM({a}.visit_by), ''), {a}.owner)
		WHEN {a}.type = 'M&E' THEN COALESCE(NULLIF(TRIM({a}.me_visit_by), ''), {a}.owner)
		WHEN {a}.type = 'Training' THEN COALESCE(NULLIF(TRIM({a}.training_entry_filled_by), ''), {a}.owner)
		WHEN {a}.type = 'Meeting' THEN COALESCE(NULLIF(TRIM({a}.mt_visit_by), ''), {a}.owner)
		ELSE COALESCE(NULLIF(TRIM({a}.visit_by), ''), {a}.owner)
	END"""


def _province_sql(alias="fv"):
	a = alias
	return f"""CASE
		WHEN {a}.type = 'Marketing' THEN COALESCE({a}.province, '')
		WHEN {a}.type = 'M&E' THEN COALESCE({a}.me_province, '')
		WHEN {a}.type = 'Training' THEN COALESCE({a}.training_province, '')
		ELSE ''
	END"""


def _category_sql(alias="fv"):
	a = alias
	return f"""COALESCE(
		NULLIF(TRIM({a}.marketing_visit_category), ''),
		NULLIF(TRIM({a}.me_activity_status), ''),
		NULLIF(TRIM({a}.training_session_category), '')
	)"""


def _sql_summary(where_clause, params):
	type_rows = frappe.db.sql(
		f"""
		SELECT IFNULL(NULLIF(TRIM(fv.type), ''), 'Other') AS type, COUNT(*) AS cnt
		FROM `tabField Visit` fv
		WHERE {where_clause}
		GROUP BY 1
		""",
		params,
		as_dict=True,
	)
	type_counts = {"Marketing": 0, "M&E": 0, "Training": 0, "Meeting": 0, "Other": 0}
	total = 0
	for r in type_rows:
		n = cint(r.cnt)
		total += n
		if r.type in type_counts:
			type_counts[r.type] += n
		else:
			type_counts["Other"] += n

	officers = frappe.db.sql(
		f"""
		SELECT DISTINCT {_officer_sql("fv")} AS officer
		FROM `tabField Visit` fv
		WHERE {where_clause}
		""",
		params,
		as_dict=True,
	)
	canonical = _staff_canonical_map()
	raw = [(r.officer or "").strip() for r in officers]
	user_names = _get_user_names(raw)
	unique = set()
	for staff_value in raw:
		mapped = user_names.get(staff_value) or staff_value or _("Unassigned")
		unique.add(_canonical_staff_label(mapped, canonical).casefold())
	active_staff = len(unique)

	return {
		"total_visits": total,
		"marketing_visits": type_counts["Marketing"],
		"me_visits": type_counts["M&E"],
		"training_visits": type_counts["Training"],
		"meeting_visits": type_counts["Meeting"],
		"other_visits": type_counts["Other"],
		"active_staff": active_staff,
		"visits_per_staff": round(total / active_staff, 1) if active_staff else 0,
	}


def _format_status(rows):
	labels = {0: "Draft", 1: "Submitted", 2: "Cancelled"}
	for row in rows:
		row["docstatus"] = labels.get(row.get("docstatus"), row.get("docstatus"))


def _build_summary(rows):
	type_counts = {"Marketing": 0, "M&E": 0, "Training": 0, "Meeting": 0, "Other": 0}
	staff_counts = {}
	raw_staff = [_get_field_staff(row) for row in rows]
	canonical = _staff_canonical_map()
	user_names = _get_user_names(raw_staff)

	for row, staff_value in zip(rows, raw_staff):
		visit_type = row.get("type") or "Other"
		if visit_type in type_counts:
			type_counts[visit_type] += 1
		else:
			type_counts["Other"] += 1

		mapped = user_names.get(staff_value) or staff_value or _("Unassigned")
		staff = _canonical_staff_label(mapped, canonical)
		staff_key = staff.casefold()
		staff_data = staff_counts.setdefault(
			staff_key,
			{
				"staff": staff,
				"total_visits": 0,
				"marketing": 0,
				"me": 0,
				"training": 0,
				"meeting": 0,
				"other": 0,
			},
		)
		staff_data["total_visits"] += 1
		if visit_type == "Marketing":
			staff_data["marketing"] += 1
		elif visit_type == "M&E":
			staff_data["me"] += 1
		elif visit_type == "Training":
			staff_data["training"] += 1
		elif visit_type == "Meeting":
			staff_data["meeting"] += 1
		else:
			staff_data["other"] += 1

	total_visits = len(rows)
	staff_wise = sorted(staff_counts.values(), key=lambda item: (-item["total_visits"], item["staff"]))
	for staff_data in staff_wise:
		staff_data["ratio"] = round(
			(staff_data["total_visits"] / total_visits * 100) if total_visits else 0,
			1,
		)

	active_staff = len(staff_wise)
	return (
		{
			"total_visits": total_visits,
			"marketing_visits": type_counts["Marketing"],
			"me_visits": type_counts["M&E"],
			"training_visits": type_counts["Training"],
			"meeting_visits": type_counts["Meeting"],
			"other_visits": type_counts["Other"],
			"active_staff": active_staff,
			"visits_per_staff": round(total_visits / active_staff, 1) if active_staff else 0,
		},
		staff_wise,
	)


def _get_field_staff(row):
	if row.get("type") == "Marketing":
		return row.get("visit_by") or row.get("owner") or _("Unassigned")
	if row.get("type") == "M&E":
		return row.get("me_visit_by") or row.get("owner") or _("Unassigned")
	if row.get("type") == "Training":
		return row.get("training_entry_filled_by") or row.get("owner") or _("Unassigned")
	if row.get("type") == "Meeting":
		return row.get("mt_visit_by") or row.get("owner") or _("Unassigned")
	return row.get("owner") or _("Unassigned")


def _staff_canonical_map():
	"""Map emails / spellings (Abdul.Kabeer) to Employee Name so one person is one staff."""
	cache = frappe.cache()
	cached = cache.get_value("tif_fsr_staff_canonical")
	if cached:
		return cached
	mapping = {}
	employees = frappe.get_all(
		"Employee",
		fields=["employee_name", "user_id"],
		limit_page_length=5000,
	)
	for emp in employees:
		name = (emp.employee_name or "").strip()
		if not name:
			continue
		mapping[name.lower()] = name
		mapping[name.replace(" ", ".").lower()] = name
		mapping[name.replace(" ", "").lower()] = name
		compact = re.sub(r"[^a-z]", "", name.lower())
		if compact:
			mapping[compact] = name
		if emp.user_id:
			mapping[emp.user_id.strip().lower()] = name
	cache.set_value("tif_fsr_staff_canonical", mapping, expires_in_sec=3600)
	return mapping


def _canonical_staff_label(raw, mapping):
	s = (raw or "").strip()
	if not s:
		return _("Unassigned")
	return (
		mapping.get(s.lower())
		or mapping.get(re.sub(r"[^a-z]", "", s.lower()))
		or s
	)


def _get_user_names(staff_values):
	emails = sorted({value for value in staff_values if value and "@" in value})
	if not emails:
		return {}
	users = frappe.get_all(
		"User",
		filters={"name": ["in", emails]},
		fields=["name", "full_name"],
	)
	return {user.name: user.full_name for user in users if user.full_name}


def _excel_value(value):
	if isinstance(value, (datetime, date, time)):
		return str(value)
	return value
